bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadFactoryBuilder.java

```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.common;
+
+import static java.util.concurrent.Executors.defaultThreadFactory;
+
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * A builder for {@link ThreadFactory} instances. This builder is intended to be use for creating thread factories to be
+ * used, e.g., when creating {@link Executor}s via the {@link Executors} utility methods.
+ * <p>
+ * The built {@link ThreadFactory} uses a wrapped {@link ThreadFactory} to create threads (defaulting to
+ * {@link Executors#defaultThreadFactory()}, and then overwrites thread properties as indicated in the build process.
+ *
+ * @author Henning Sudbrock - initial contribution
+ */
+@NonNullByDefault
+public class ThreadFactoryBuilder {
+
+    private @Nullable ThreadFactory wrappedThreadFactory;
+    private @Nullable String namePrefix;
+    private boolean daemonThreads;
+    private @Nullable UncaughtExceptionHandler uncaughtExceptionHandler;
+    private @Nullable Integer priority;
+
+    /**
+     * Creates a fresh {@link ThreadFactoryBuilder}.
+     *
+     * @return A {@link ThreadFactoryBuilder}
+     */
+    public static ThreadFactoryBuilder create() {
+        return new ThreadFactoryBuilder();
+    }
+
+    private ThreadFactoryBuilder() {
+        // use static factory method to create ThreadFactoryBuilder
+    }
+
+    /**
+     * Sets the wrapped thread factory used to create threads. If set to null, {@link Executors#defaultThreadFactory()}
+     * is used. Defaults to null.
+     *
+     * @param wrappedThreadFactory the wrapped thread factory to be used
+     * @return this {@link ThreadFactoryBuilder} instance
+     */
+    public ThreadFactoryBuilder withWrappedThreadFactory(@Nullable ThreadFactory wrappedThreadFactory) {
+        this.wrappedThreadFactory = wrappedThreadFactory;
+        return this;
+    }
+
+    /**
+     * Sets the name prefix to be used by the {@link ThreadFactory}.
+     * <p>
+     * The threads created by the {@link ThreadFactory} will be named 'ESH-namePrefix-i', where i is an integer
```

Comment:
I personally feel it is a good idea to use the `ThreadFactoryBuilder` to customize the thread factories. I am a bit concerned about the thread factory naming pattern since it follows a naming pattern of `ESH-NamePrefix-i` where i is an integer. I believe it would be unmanageable in the future if we name all thread factories to have a prefix of `ESH`. Even though it wouldn't do any harm, I believe it would have been nicer not to specify such fixed prefix. It makes more sense to follow the factory naming pattern to `ESH-NamePrefix-i` if and only if the consumers use ESH's `ThreadPoolManager`. On the other hand, if java's in-built executors are used, we could remove such `ESH` prefix restriction.

Reply:
As I mentioned, there won't be any problem but it is just a matter of practice. If I just use java's simple executor and an utility class (`ThreadFactoryBuilder`) from ESH, I would not expect my thread names to have ESH prefix. I would expect it to happen if I use ESH's `ThreadPoolManager` instead. Since `ThreadFactoryBuilder` will not be managing the actual threads, it would make more sense not to have "ESH" as prefix. If I use the `ExecutorService`, ESH core will not even manage my threads. And in this case, it would be more restrictive to have the thread pools prefixed with "ESH".

Reply:
I think that the new `ThreadFactoryBuilder` might be valuable for solutions as well, and not only for use in ESH. And for developers of solutions, it would be good if they could select a different prefix for their threads, as their threads do not come from the ESH codebase - hence I agree with @amitjoy.

Of course, those solutions could provide their own `ThreadFactoryBuilder`, but I would prefer if they could easily use the one that is included with ESH.

Suggestion: Make it possible in the builder to set the prefix, with JavaDoc indicating that this should be used within solutions, but not within ESH code (so that threads from the ESH framework share the same prefix). The default prefix would be `"ESH"`. With this approach, solutions wouldn't be completely free in selecting the name (compared to, e.g., Guava's ThreadFactory builder, where a format string can be provided that may contain a parameter for the counter `i`), but it should be sufficient for easy reuse in solutions. I would not provide complete freedom, so that the builder stays easily usable from within ESH.

wdyt?

Reply:
Okay for me.

Reply:
I did not get to change with @amitjoy's suggestion; will hopefully add a commit soon...